clients/hadoopfs/src/main/java/io/lakefs/FileSystemTracer.java

```diff
@@ -0,0 +1,353 @@
+
+    public static final Logger LOG = LoggerFactory.getLogger(FileSystemTracer.class);
+    private static final String TRACER_WORKING_DIR = "fs.lakefs.tracer.working.dir";
+    private static final String USE_LAKEFS_RESPONSE = "fs.lakefs.tracer.use.lakefs";
+    private static final Object S3_URI_PREFIX = "s3";
+    private static final Object RESULTS_COMPARISON = "[RESULTS_COMPARISON]";
+
+
+    /*
+     A property that determines which file system's response the FileSystemTracer returns.
+     */
+    private boolean useLakeFSFileSystemRes;
+    private LakeFSFileSystem lfsFileSystem;
+    private FileSystem s3AFileSystem;
+    private String s3aPathPrefix;
+    private String lfsPathPrefix;
+
+    /**
+     * Transforms a lakefs path into an s3 path.
+     * Example:
+     *   in: lakefs://repository/branch/key=1/a.parquet
+     *   out: s3://${fs.lakefs.tracer.working.dir}/key=1/a.parquet
+     */
+    private Path translateLakeFSPathToS3APath(Path path) {
+        ObjectLocation objectLoc = lfsFileSystem.pathToObjectLocation(path);
+        String lakefsPath = path.toString();
+        String lakefsPrefix = String.format("%s://%s/%s", objectLoc.getScheme(), objectLoc.getRepository(),
+                objectLoc.getRef());
+
+        String s3aPath = lakefsPath.replace(lakefsPrefix, s3aPathPrefix);
+        LOG.trace("Converted {} to {}", path, s3aPath);
+
+        return new Path(s3aPath);
+    }
+
+    /**
+     * Transforms an S3A path into a lakefs path.
+     * This method is used to transform the output of S3AFileSystem operations that include paths, so that Spark
+     * continues to direct it's requests to the FileSystemTracer that handles paths with the "lakefs://" prefix.
+     * Example:
+     *   in: s3://${fs.lakefs.tracer.working.dir}/key=1/a.parquet
+     *   out: lakefs://repository/branch/key=1/a.parquet
+     */
+    private Path translateS3APathToLakeFSPath(Path path) {
+        String p = path.toString();
+        String lfsPath = p.replace(s3aPathPrefix, lfsPathPrefix);
+        LOG.trace("Converted {} to {}", path, lfsPath);
+        return new Path(lfsPath);
+    }
+
+
+    @Override
+    public void initialize(URI name, Configuration conf) throws IOException {
+        lfsFileSystem = new LakeFSFileSystem();
+        lfsFileSystem.initialize(name, conf);
+        ObjectLocation loc = lfsFileSystem.pathToObjectLocation(new Path(name));
+        lfsPathPrefix = String.format("%s://%s/%s", loc.getScheme(), loc.getRepository(), loc.getRef());
+
+        String tracerWorkingDir = conf.get(TRACER_WORKING_DIR);
+        s3aPathPrefix = String.format("%s://%s", S3_URI_PREFIX, tracerWorkingDir);
+        Path s3aPath = new Path(s3aPathPrefix);
+        s3AFileSystem = s3aPath.getFileSystem(conf);
+
+        useLakeFSFileSystemRes = conf.getBoolean(USE_LAKEFS_RESPONSE, false);
+        LOG.trace("Initialized FileSystemTracer, fs.lakefs.tracer.use.lakefs: {}", useLakeFSFileSystemRes);
```

Comment:
LOG log everything related to FileSystemTracer so no need to specify it also in the log messages